tensorboard/db.py

```diff
@@ -387,6 +411,333 @@ def _cursor(self):
+    raise NotImplementedError('Cursor.nextset not supported')
+
+  @property
+  def description(self):
+    """Returns information about each column in result set.
+
+    :rtype: list[tuple[str, int, int, int, int, int bool]]
+    """
+    self._check_that_read_query_was_issued()
+    return self._cursor.description
+
+  @property
+  def rowcount(self):
+    """Returns number of rows retrieved by last read query.
+
+    :rtype: int
+    """
+    self._check_that_read_query_was_issued()
+    return self._cursor.rowcount
+
+  @property
+  def lastrowid(self):
+    """Raises NotImplementedError.
+
+    This is not possible to implement in TensorBoard's transaction
+    system, which defers writes to the end.
+
+    :rtype: int
+    """
+    self._check_that_read_query_was_issued()
+    return self._cursor.rowcount
+
+  def _get_arraysize(self):
+    self._init_cursor()
+    return self._cursor.arraysize
+
+  def _set_arraysize(self, arraysize):
+    self._init_cursor()
+    self._cursor.arraysize = arraysize
+
+  arraysize = property(_get_arraysize, _set_arraysize)
+
+  def close(self):
+    """Closes resources associated with cursor."""
+    if self._cursor is not None:
+      self._cursor.close()
+      self._cursor = None
+    self._is_closed = True
+
+  def __iter__(self):
+    """Returns iterator over results of last read query.
+
+    :rtype: types.GeneratorType[tuple[object]]
+    """
+    self._check_that_read_query_was_issued()
+    for row in self._cursor:
+      yield row
+
+  def callproc(self, procname, parameters=()):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.callproc not supported')
+
+  def setinputsizes(self, sizes):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.setinputsizes not supported')
+
+  def setoutputsize(self, size, column):
+    """Raises NotImplementedError."""
+    raise NotImplementedError('Cursor.setoutputsize not supported')
+
+  def _init_cursor(self):
+    self._check_closed()
+    if self._cursor is None:
+      self._cursor = self._actual_connection.cursor()
+
+  def _check_that_read_query_was_issued(self):
+    self._check_closed()
+    if self._cursor is None:
+      raise ValueError('no read query was issued')
+
+  def _check_closed(self):
+    if self._is_closed:
+      raise ValueError('cursor was closed')
+
+
+class _DeferredCursor(object):
+  """Helper class for write queries."""
+
+  def __init__(self):
+    self._queries = []  # type: list[tuple[str, str, object]]
+
+  def execute(self, sql, parameters):
+    """Executes query in the future.
+
+    :type sql: str
+    :type parameters: tuple[object]
+    """
+    self._queries.append(('execute', sql, parameters))
+
+  def executemany(self, sql, seq_of_parameters):
+    """Executes multi-query in the future.
+
+    :type sql: str
+    :type seq_of_parameters: list[tuple[object]]
+    """
+    self._queries.append(('executemany', sql, seq_of_parameters))
+
+  def replay(self, connection):
```

Comment:
Why is this called "replay"? The name doesn't communicate to me that it executes deferred writes.